velox/docs/functions/spark/array.rst

```diff
@@ -125,13 +125,34 @@ Array Functions
 .. spark:function:: array_sort(array(E)) -> array(E)
 
     Returns an array which has the sorted order of the input array(E). The elements of array(E) must
-    be orderable. Null elements will be placed at the end of the returned array. ::
+    be orderable. Null/NaN elements will be placed at the end of the returned array. ::
 
         SELECT array_sort(array(1, 2, 3)); -- [1, 2, 3]
         SELECT array_sort(array(3, 2, 1)); -- [1, 2, 3]
         SELECT array_sort(array(2, 1, NULL); -- [1, 2, NULL]
         SELECT array_sort(array(NULL, 1, NULL)); -- [1, NULL, NULL]
         SELECT array_sort(array(NULL, 2, 1)); -- [1, 2, NULL]
+        SELECT array_sort(array(4.0, float('nan'), 3.0)); -- [3.0, 4.0, NaN]
+        SELECT array_sort(array(array(), array(1, 3, NULL), array(NULL, 6), NULL, array(2,1))) -- [[], [NULL, 6], [1, 3, NULL], [2, 1], NULL]
+
+.. spark:function:: array_sort(array(T), function(T,U)) -> array(T)
+    :noindex:
+
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.  ::
```

Comment:
NULL or NaN for floating type?

Reply:
Currently for lambda function returned values, Nan is not handled. Do we need to handle `NaN` since it shouldn't returned by lambda functions

Reply:
Oh, sorry, I overlooked this before, for array_sort with lambda functions, it supports sorting with `NaN` in `SimpleVector.comparePrimitiveAsc` and follows the logic of `NaN is before NULL`, I also add a test to cover this.

Reply:
Let me explain more here. After looking into the presto/spark implementation, they both say

> It returns -1, 0, or 1 as the first nullable element is less than, equal to, or greater than the second nullable element. If the comparator function returns other values (including NULL), the query will fail and raise an error

see [presto](https://prestodb.io/docs/current/functions/array.html#array_sort) and [spark](https://docs.databricks.com/en/sql/language-manual/functions/array_sort.html) (Though spark says it doesn't support returning null values, but it doesn't throw errors for query like `SELECT array_sort(ARRAY ('bc', 'ab', 'dc'), (x, y) -> IF(x < y, 1, IF(x = y, 0, null)))` in Spark3.2 which might be a bug)

So null values and NaN shouldn't be return for lambda function `function(T,T, int)`, and in `SimpleComparisonMatcher`, we do the match that the return value must be int.

`SimpleComparisonMatcher` could optimize `function(T,T, int)` to `function(T, U)` where `U` is orderable(not limited to int), it's possible that it creates float values, such as `function(float, float, int)`: `IF( x > y, 1, IF(x < y, -1, 0))` will be optimized to `function(float, float)`: `x -> x`, at such point, they should still be the same since both goes into `SimpleVector.compare` to do the comparison(except NULLs are filtered in `ArraySort.sortElements` in advance to respect `nullsFirst` flag). And inside `SimpleVector.compare`, `NaN` is smaller than `Null`.

Reply:
I tried `SELECT array_sort(ARRAY ('bc', 'ab', 'dc'), (x, y) -> IF(x < y, 1, IF(x = y, 0, null)))` in Spark 3.5 and got below exception. Would you like the add a unit test for this case to make sure exception is thrown?
> Caused by: org.apache.spark.SparkException: [COMPARATOR_RETURNS_NULL] The comparator has returned a NULL for a comparison between dc and dc. It should return a positive integer for "greater than", 0 for "equal" and a negative integer for "less than". To revert to deprecated behavior where NULL is treated as 0 (equal), you must set "spark.sql.legacy.allowNullComparisonResultInArraySort" to "true".

I also notice Spark requires the function must return integer type, would you like to confirm?
https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala#L412-L421